### PR TITLE
Refactor FXIOS-8560 [v124] Missing original "Show Search Suggestions" on Search Settings

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8560)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/19030)

## :bulb: Description

- Restore `Show Search Suggestions` option for `Default Search Engines` section when the experiment is off.
- Change logic in order that when one of the options `Search Bookmarks`, `Search Browsing History`, `Search Sync Tabs` is on, the `Show in Private Sessions` option is available and vice versa when all these options are off, hide `Show in Private Sessions` from `Address Bar - Firiefox Suggest` section.
- Remove `Show in Private Sessions` option from `Suggestions from Search Engines` section for the `Firefox Suggest` feature.
## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 1]
```diff
@@ -145,15 +146,31 @@ final class SearchSettingsTableViewController: ThemedTableViewController, Featur
...
+                        prefKey: PrefsKeys.SearchSettings.showSearchSuggestions,
+                        defaultValue: model.shouldShowSearchSuggestions,
+                        titleText: String.localizedStringWithFormat(
+                            .Settings.Search.ShowSearchSuggestions
+                        ),
+                        cell: cell,
+                        selector: #selector(didToggleSearchSuggestions)
+                    )
+                }
+            default: break
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift:172]: I believe we want to move this section down with the private toggle and should be enabled for all cases.

**Review by cyndichin — CHANGES_REQUESTED** [review]
thanks for the update @PARAIPAN9! A couple of changes requested based on the experience we want to provide, let me know if you have any questions!

File: firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift [hunk 2]
```diff
@@ -2046,16 +2046,22 @@ class BrowserViewController: UIViewController,
...
+        let isSettingEnabled = if firefoxSuggestFlagEnabled {
+            !profile.searchEngines.isPrivateModeSettingEnabled &&
+            tabManager.selectedTab?.isPrivate == true
+        } else {
+            feltPrivacyFlagEnabled &&
+            !alwaysShowSearchSuggestionsView &&
+            !profile.searchEngines.shouldShowPrivateModeSearchSuggestions
+        }
+
+        return isSettingEnabled
```
[the remark this answers is not in the document]
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:2064]: Confirmed with Andy, that the private toggle setting under FX Suggest should only hide the section in search view and not display dimming view.

**Comment by isabelrios** [comment]
@PARAIPAN9 looks like there is a UI test failing 3 out of 3 a few times, it may be that the changes in this PR are chaging its behaviour.. 

`/Users/vagrant/git/firefox-ios/firefox-ios-tests/Tests/XCUITests/URLValidationTests.swift:18 - Failed to tap "FirefoxSuggestShowNonSponsoredSuggestions" Switch: No matches found for Descendants matching type Table from input`

Please let us know if you need help with it to be green again..

File: firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 3]
```diff
@@ -145,15 +142,16 @@ final class SearchSettingsTableViewController: ThemedTableViewController, Featur
...
+                engine = model.defaultEngine
+                cell.editingAccessoryType = .disclosureIndicator
+                cell.accessibilityLabel = .Settings.Search.AccessibilityLabels.DefaultSearchEngine
+                cell.accessibilityValue = engine.shortName
+                cell.textLabel?.text = engine.shortName
+                cell.imageView?.image = engine.image.createScaled(IconSize)
+                cell.imageView?.layer.cornerRadius = 4
+                cell.imageView?.layer.masksToBounds = true
+                cell.applyTheme(theme: themeManager.currentTheme)
+
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift:154]: nit: extra indention, inconsistent with the other cases.

File: firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 4]
```diff
@@ -345,7 +342,9 @@ final class SearchSettingsTableViewController: ThemedTableViewController, Featur
             // But the option to add a Search Engine is.
             return model.orderedEngines.count
         case .searchEnginesSuggestions:
-            return SearchSuggestItem.allCases.count
+            return featureFlags.isFeatureEnabled(.feltPrivacyFeltDeletion, checking: .buildAndUser) &&
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift:345]:
```suggestion
            return featureFlags.isFeatureEnabled(.feltPrivacySimplifiedUI, checking: .buildOnly) &&
```

File: firefox-ios/Client/Frontend/Browser/SearchViewController.swift [hunk 5]
```diff
@@ -1089,9 +1130,7 @@ class SearchViewController: SiteTableViewController,
         case SearchListSection.remoteTabs.rawValue:
             return hasFirefoxSuggestions
         case SearchListSection.searchSuggestions.rawValue:
-            return viewModel.isPrivate ?
-            model.shouldShowPrivateModeSearchSuggestions :
-            model.shouldShowSearchSuggestions
+            return shoudShowSearchEngineSuggestions
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/SearchViewController.swift:1133]:
```suggestion
            return shouldShowSearchEngineSuggestions
```

File: firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift [hunk 6]
```diff
@@ -192,12 +192,15 @@ class SearchEngines {
...
+    }
+
     var isPrivateModeSettingEnabled: Bool {
-        return shouldShowPrivateModeSearchSuggestions ||
-               shouldShowPrivateModeFirefoxSuggestions ||
-               shouldShowBookmarksSuggestions ||
-               shouldShowSyncedTabsSuggestions ||
-               shouldShowBrowsingHistorySuggestions
+        return shouldShowPrivateModeFirefoxSuggestions ||
+               shouldShowSearchSuggestions
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift:203]: looks like we can remove this now.

File: firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift [hunk 7]
```diff
@@ -192,12 +192,15 @@ class SearchEngines {
         }
     }
 
+    var shouldHidePrivateModeFirefoxSuggestSetting: Bool {
+        return !shouldShowBookmarksSuggestions &&
+        !shouldShowSyncedTabsSuggestions &&
+        !shouldShowBrowsingHistorySuggestions
+    }
+
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift:200]: This logic seems to be related to Firefox Suggest, but lives in Search Engine class (which imo is separate). The `SearchSettingsTableViewController` seems to be a bit large, do we see a benefit in creating a viewmodel to hold all these boolean values instead? Just consideration, not a blocker on this PR.

File: firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift [hunk 8]
```diff
@@ -559,6 +555,48 @@ final class SearchSettingsTableViewController: ThemedTableViewController, Featur
         cell.selectionStyle = .none
     }
 
+    private func didToggleFirefoxSuggestions(
+        _ toggle: ThemedSwitch,
+        suggestionType: FirefoxSuggestItem
+    ) {
+        var shouldShowSuggestions = false
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Settings/SearchSettingsTableViewController.swift:562]:
```suggestion
        var shouldShowPrivateSuggestionsSetting = false
```
Took me a while to understand what we are doing, but it seems we want to show or hide the private suggestions setting based on the three settings. Maybe it'll help to write a comment or rename?

File: firefox-ios/Client/Frontend/Browser/SearchViewController.swift [hunk 9]
```diff
@@ -977,8 +1017,9 @@ class SearchViewController: SiteTableViewController,
             if !featureFlags.isFeatureEnabled(.firefoxSuggestFeature,
                                               checking: .buildAndUser) {
                 model.shouldShowSyncedTabsSuggestions = true
+                model.shouldShowPrivateModeFirefoxSuggestions = true
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/SearchViewController.swift:1020]: curious - why do we set this to true when the feature flag is disabled?

**Review by cyndichin — COMMENTED** [review]
thanks for the updates! this is a tricky one as there are lot of cases to consider - have a couple of comments but only one blocking change regarding the feature flag.

File: firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift (hunk 7 above)
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/Browser/SearchEngines/SearchEngines.swift]: I'll relocate the logic to the SearchSettingsTableViewController, and I'll establish the view model in another PR, migrating all necessary logic there. Time constraints are necessitating us to accelerate this process.

File: firefox-ios/Client/Frontend/Browser/SearchViewController.swift (hunk 9 above)
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/Browser/SearchViewController.swift:1020]: This is to ensure their accuracy, as when switching experiments, Firefox Suggest might leave them in a disabled state resulting in being hidden in the other experiment.

**Review by cyndichin — APPROVED** [review]
thanks for the updates!

**Comment by cyndichin** [comment]
@mergify release/v124

**Comment by cyndichin** [comment]
@mergify backport release/v124

**Pull request merged**